src/angular/components/change-password.component.ts

```diff
@@ -0,0 +1,263 @@
+            this.platformUtilsService.showToast('error', this.i18nService.t('errorOccurred'),
+                this.i18nService.t('masterPassRequired'));
+            return;
+        }
+        if (this.newMasterPassword.length < 8) {
+            this.platformUtilsService.showToast('error', this.i18nService.t('errorOccurred'),
+                this.i18nService.t('masterPassLength'));
+            return;
+        }
+        if (this.newMasterPassword !== this.confirmNewMasterPassword) {
+            this.platformUtilsService.showToast('error', this.i18nService.t('errorOccurred'),
+                this.i18nService.t('masterPassDoesntMatch'));
+            return;
+        }
+
+        const strengthResult = this.passwordGenerationService.passwordStrength(this.newMasterPassword,
+            this.getPasswordStrengthUserInput());
+
+        if (this.enforcedPolicyOptions != null &&
+            !this.policyService.evaluateMasterPassword(
+                strengthResult.score,
+                this.newMasterPassword,
+                this.enforcedPolicyOptions)) {
+            this.platformUtilsService.showToast('error', this.i18nService.t('errorOccurred'),
+                this.i18nService.t('masterPasswordPolicyRequirementsNotMet'));
+            return;
+        }
+
+        if (strengthResult != null && strengthResult.score < 3) {
+            const result = await this.platformUtilsService.showDialog(this.i18nService.t('weakMasterPasswordDesc'),
+                this.i18nService.t('weakMasterPassword'), this.i18nService.t('yes'), this.i18nService.t('no'),
+                'warning');
+            if (!result) {
+                return;
+            }
+        }
+
+        if (this.rotateEncKey) {
+            await this.syncService.fullSync(true);
+        }
+
+        const email = await this.userService.getEmail();
+        const kdf = await this.userService.getKdf();
+        const kdfIterations = await this.userService.getKdfIterations();
+        const newKey = await this.cryptoService.makeKey(this.newMasterPassword, email.trim().toLowerCase(),
+            kdf, kdfIterations);
+        const newMasterPasswordHash = await this.cryptoService.hashPassword(this.newMasterPassword, newKey);
+        const newEncKey = await this.cryptoService.remakeEncKey(newKey);
+
+        if (this.isChangePasswordNoCompare) {
+            const noCompareRequest = new PasswordNoCompareRequest();
+            noCompareRequest.newMasterPasswordHash = newMasterPasswordHash;
+            noCompareRequest.key = newEncKey[1].encryptedString;
+
+            try {
+                this.formPromise = this.apiService.postPasswordNoCompare(noCompareRequest);
+                await this.formPromise;
+
+                if (this.onSuccessfulChangePassword != null) {
+                    this.onSuccessfulChangePassword();
+                } else {
+                    this.router.navigate([this.successRoute]);
+                }
+            } catch { }
```

Comment:
What happens when there is a failure? Should the user be notified?

Reply:
Added default error message toast.